### PR TITLE
Write hie info

### PR DESCRIPTION
**Pull request by balacij**
Contributes to #3700 by writing HIE information necessary for both `stan` and `weeder`

### CODE REVIEW

**Comment by balacij** [comment]
Force-pushed to remove a commit I didn't intend to write to this branch...

File: code/Makefile [hunk 1]
```diff
@@ -492,16 +492,17 @@ deploy: analysis graphs convertAnalyzed packagedeps docs tex doxygen ##@Deploy G
 
 # Removes all generated folders.
 $(filter $(CLEAN_GF_PREFIX)%, $(CLEAN_FOLDERS)): $(CLEAN_GF_PREFIX)%:
-	- rm -rf "./$*/"
+	rm -rf "./$*/"
```
Comment (JacquesCarette) [inline code/Makefile:495]: why this change?

Comment (balacij) [reply code/Makefile:495]: It's _almost completely_ redundant. Passing `-f` `rm` already suppresses the majority of errors, which is what `-` would do for that Makefile line. The errors that `rm -f` return all have to do with a system that we somehow were trying to edit, but which we were stopped by the OS for some reason (e.g., we accidentally tried to delete System32 or any other read-only file, the local working folders became read-only [permission loss], etc.). I think we would want to see those errors, because it would mean something has gone very seriously wrong.

**Comment by balacij** [comment]
Yep, [`.hie` files](https://gitlab.haskell.org/ghc/ghc/-/wikis/hie-files) (and the directory) contain compiler-related information that was found while GHC was working with it. From what I've understood, its ultimate goal is to enable external tools to 'search/analyze/manipulate the AST' as if it were built into the compiler itself, so it has an easier time dealing with the AST //instead of needing to work with the CST. The [`hiedb` package](https://hackage.haskell.org/package/hiedb) allows us to write Haskell code that handles the `.hie` information (hence enumerating what information we can readily see from the GHC output), and it also has a nifty ['Query'](https://hackage.haskell.org/package/hiedb-0.6.0.0/docs/HieDb-Query.html) that we might be able to use as well! For example, we have those two scripts in the `scripts/` folder -- I forget what they do completely, but I know they scan the source code -- that could probably be simplified by relying on the `.hie` information instead of a string search. Maybe we make that a work task actually? I'm not sure how robust those scripts are.

**Comment by balacij** [comment]
@JacquesCarette I see on #3739, you mentioned that you wanted more information about this -- did the above comment clarify everything?

**Comment by JacquesCarette** [comment]
Upon re-reading: yes.
